cmd/snap/cmd_run.go

```diff
@@ -1013,3 +1062,256 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
+	return strings.TrimSpace(string(uuidBytes)), err
+}
+
+var osGetpid = os.Getpid
+
+var cgroupProcessPathInTrackingCgroup = cgroup.ProcessPathInTrackingCgroup
+
+var errCannotTrackProcess = errors.New("cannot track application process")
+
+var createTransientScope = func(securityTag string) error {
+	if !features.RefreshAppAwareness.IsEnabled() {
+		return nil
+	}
+	logger.Debugf("creating transient scope %s", securityTag)
+
+	// We ask the kernel for a random UUID. We need one because each transient
+	// scope needs a unique name. The unique name is comprosed of said UUID and
+	// the snap security tag.
+	uuid, err := randomUUID()
+	if err != nil {
+		return err
+	}
+
+	// Enforcing uniqueness is preferred to reusing an existing scope for
+	// simplicity since doing otherwise and joining an existing scope has
+	// limitations:
+	// - the originally started scope must be marked as a delegate, with all consequences.
+	// - the method AttachProcessesToUnit is unavailable on Ubuntu 16.04
+	//
+	// TODO: invert the arguments so that security tag comes first, UUID
+	// second. This will result in nicer scope names but needs to be
+	// coordinated with code in sandbox/cgroup and with the spread tests
+	// exercising that.
+	unitName := fmt.Sprintf("snap.%s.%s.scope", uuid, strings.TrimPrefix(securityTag, "snap."))
+	pid := osGetpid()
+
+	// The scope is created with a DBus call to systemd running either on
+	// system or session bus. We have a preference for session bus, as this is
+	// where applications normally go to. When a session bus is not available
+	// and the invoking user is root, we use the system bus instead.
+	//
+	// It is worth noting that hooks will not normally have a session bus to
+	// connect to, as they are invoked as descendants of snapd, and snapd is a
+	// service running outside of any session.
+	isSessionBus := true
+	conn, err := dbusSessionBus()
+	if err != nil {
+		logger.Debugf("session bus is not available: %s", err)
+		if osGetuid() == 0 {
+			logger.Debugf("falling back to system bus")
+			isSessionBus = false
+			conn, err = dbusSystemBus()
+			if err != nil {
+				logger.Debugf("system bus is not available: %s", err)
+			} else {
+				logger.Debugf("using system bus now, session bus was not available")
+			}
+		}
+	} else {
+		logger.Debugf("using session bus")
+	}
+
+	// Session or system bus might be unavailable. To avoid being fragile
+	// ignore all errors when establishing session bus connection to avoid
+	// breaking user interactions. This is consistent with similar failure
+	// modes below, where other parts of the stack fail.
+	//
+	// Ideally we would check for a distinct error type but this is just an
+	// errors.New() in go-dbus code.
+	if err != nil {
+		return errCannotTrackProcess
+	}
+
+tryAgain:
+	// Create a transient scope by talking to systemd over DBus.
+	if err := doCreateTransientScope(conn, unitName, pid); err != nil {
+		switch err {
```

Comment:
same for the switch and its stanzas here

Reply:
Those are now all tested.